File: contracts/manifest.py
"""
Audioviz Manifest Schema (Python contract)
Matches renderer/src/types/manifest.ts 1:1
"""
from dataclasses import dataclass, field, asdict
from typing import List, Dict, Any, Optional, Literal

VisualizerPresetId = Literal["sphere", "iris", "neural", "monolith"]


@dataclass
class AudioMultibandFeatures:
    bass: List[float]  # 30 FPS normalized [0.0, 1.0]
    mids: List[float]  # 30 FPS normalized [0.0, 1.0]
    treble: List[float]  # 30 FPS normalized [0.0, 1.0]
    transients: List[int]  # Frame indices where transient drum kicks or drops occur
    bpm: float = 120.0  # Detected musical tempo
    beatFrames: List[int] = field(default_factory=list)  # Exact frame indices of quarter-note beats
    downbeatFrames: List[int] = field(default_factory=list)  # 1st-beat-of-bar downbeats
    vocalEnergy: List[float] = field(default_factory=list)  # 30 FPS isolated vocal envelope
    macroEnergy: List[float] = field(default_factory=list)  # 30 FPS structural build-up/drop curve
    musicalKey: str = ""  # Harmonic key tonality (e.g. "D Minor")


@dataclass
class LyricLine:
    text: str = ""
    startFrame: int = 0
    endFrame: int = 0
    isHero: bool = False


@dataclass
class PresetConfig:
    id: VisualizerPresetId
    version: str = "1.0.0"
    parameters: Dict[str, Any] = field(default_factory=dict)


@dataclass
class VideoConfig:
    width: int = 1920
    height: int = 1080
    fpsNumerator: int = 30
    fpsDenominator: int = 1
    frameCount: int = 300


@dataclass
class AudioConfig:
    masterUri: str
    sha256: str
    sampleRate: int = 44100
    vocalStemUri: Optional[str] = None
    bassStemUri: Optional[str] = None
    features: Optional[AudioMultibandFeatures] = None


@dataclass
class LyricsConfig:
    lines: List[LyricLine] = field(default_factory=list)


@dataclass
class RenderManifest:
    schemaVersion: int
    jobId: str
    seed: int
    preset: PresetConfig
    video: VideoConfig
    audio: AudioConfig
    lyrics: LyricsConfig
    environment: Dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)

File: renderer/src/types/manifest.ts
/**
 * Core Data Contracts for Stem-Aware 3D Audio Visualizer (Audioviz)
 * Incorporates Astra Architectural Guardrails:
 * - Deterministic frame evaluation: timeSeconds = frame * fpsDenominator / fpsNumerator
 * - Immutable Pre-Render Manifest (100% precalculated features before Remotion frame capture)
 */

export type VisualizerPresetId = "sphere" | "iris" | "neural" | "monolith";

export interface AudioMultibandFeatures {
  /** 30 FPS normalized bass energy array [0.0, 1.0] */
  bass: number[];
  /** 30 FPS normalized mid-frequency energy array [0.0, 1.0] */
  mids: number[];
  /** 30 FPS normalized high/treble energy array [0.0, 1.0] */
  treble: number[];
  /** Frame indices where transient drum kicks or drops occur */
  transients: number[];
  /** Detected musical tempo in BPM */
  bpm?: number;
  /** Exact frame indices of quarter-note beats */
  beatFrames?: number[];
  /** 1st-beat-of-bar downbeats */
  downbeatFrames?: number[];
  /** 30 FPS isolated vocal envelope */
  vocalEnergy?: number[];
  /** 30 FPS isolated drums envelope */
  drumsEnergy?: number[];
  /** 30 FPS structural build-up/drop curve */
  macroEnergy?: number[];
  /** Harmonic key tonality (e.g. "D Minor") */
  musicalKey?: string;
}

export interface LyricLine {
  text: string;
  startFrame: number;
  endFrame: number;
  isHero?: boolean;
}

export interface RenderManifest {
  schemaVersion: 1;
  jobId: string;
  seed: number;
  preset: {
    id: VisualizerPresetId;
    version: string;
    parameters: Record<string, number | string | boolean>;
  };
  video: {
    width: 1080 | 1920;
    height: 1080;
    fpsNumerator: 30;
    fpsDenominator: 1;
    frameCount: number;
  };
  audio: {
    masterUri: string;
    vocalStemUri?: string | null;
    bassStemUri?: string | null;
    sha256: string;
    sampleRate: number;
    features: AudioMultibandFeatures;
  };
  lyrics: {
    lines: LyricLine[];
  };
  environment?: {
    containerDigest?: string;
    threeVersion?: string;
    remotionVersion?: string;
    gpuBackend?: string;
  };
}

export interface VisualReferenceContract {
  presetId: VisualizerPresetId;
  presetVersion: string;
  camera: {
    projection: "perspective" | "orthographic";
    fov: number;
    position: [number, number, number];
    target: [number, number, number];
  };
  composition: {
    width: number;
    height: number;
    fps: number;
    colorSpace: "linear-srgb";
    toneMapping: "ACESFilmic";
    safePaddingPx: number; // 160px safe margin
  };
  goldenFrames: {
    calmTimestampSeconds: number;
    dropTimestampSeconds: number;
    ssimTolerance: number; // >= 0.92
  };
}
